### PR TITLE
Stateless components

### PR DESCRIPTION
**Pull request by iamdustan**
Detect and wrap stateless function components.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -88,13 +89,19 @@ export default function ({ Plugin, types: t }) {
...
+   * Infers a displayName from either a class node, createClass() call, or a stateless component variable declaration
    */
-  function findDisplayName(node) {
+  function findDisplayName(node, parent) {
     if (node.id) {
       return node.id.name;
     }
+
     if (!node.arguments) {
+      // likely stateless function component. Check if assigned to a variable
```
Comment (iamdustan) [inline src/index.js]: threading the `parent` node through to here so that the following example resolves a displayName of `MyComponent`

```
const MyComponent = (props) => <div />
```

File: src/index.js [hunk 2]
```diff
@@ -273,8 +280,60 @@ export default function ({ Plugin, types: t }) {
...
+        }
+
+        if (this.state.isPotentialStatelessComponent) {
+          delete this.state.isPotentialStatelessComponent;
+          this.state.isStatelessComponent = true;
+        }
+      },
+
+      [statelessVisitorTypes]: {
+        enter(node, parent, scope, file) {
```
Comment (iamdustan) [inline src/index.js]: I’m opting to do do some simple internal state tracking on enter to mark as a StatelessComponent and then return the wrapped function on exit.

File: test/fixtures/stateless/expected.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  _$ComponentInsideFunction: {
+    displayName: "ComponentInsideFunction",
+    isInFunction: true
+  },
+  _$Unknown3: {
+    isInFunction: true
+  },
+  _$factory: {
+    displayName: "factory",
+    isInFunction: true
```
Comment (iamdustan) [inline test/fixtures/stateless/expected.js]: I’m not certain if all of these are actually correct; I suspect not, though.

Comment (iamdustan) [inline test/fixtures/stateless/expected.js]: e.g.

```
const MyComp = (props) => <div />;
```

probably should not be marked as `isInFunction`, correct?

**Comment by iamdustan** [comment]
Oh shucks, I just saw #34 and the reference to react-proxy@react-0.14. Whoops :)

**Comment by joshblack** [comment]
@iamdustan if you have time, would love some review on some of the changes I made! Have just been doing what I can to make the test cases pass or generate new ones for edge cases.

**Comment by iamdustan** [comment]
:) @joshblack I’ve just been adding support for these things to another project. reactjs/react-docgen#28. Divide and conquer!

**Comment by iamdustan** [comment]
I’m kind of biased towards not supporting edge cases and guiding folks down the preferred path.
